pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1107 @@
+
+def _before_indexed_interval_var(visitor, child):
+    cpx_vars = {}
+    for i, v in child.items():
+        cpx_interval_var = _get_docplex_interval_var(visitor, v)
+        visitor.var_map[id(v)] = cpx_interval_var
+        visitor.pyomo_to_docplex[v] = cpx_interval_var
+        cpx_vars[i] = cpx_interval_var
+    return False, (_GENERAL, cpx_vars)
+
+def _before_interval_var_start_time(visitor, child):
+    _id = id(child)
+    interval_var = child.get_associated_interval_var()
+    if _id not in visitor.var_map:
+        cpx_interval_var = _get_docplex_interval_var(visitor, interval_var)
+
+    return False, (_START_TIME, visitor.var_map[id(interval_var)])
+
+def _before_interval_var_end_time(visitor, child):
+    _id = id(child)
+    interval_var = child.get_associated_interval_var()
+    if _id not in visitor.var_map:
+        cpx_interval_var = _get_docplex_interval_var(visitor, interval_var)
+
+    return False, (_END_TIME, visitor.var_map[id(interval_var)])
+
+def _before_interval_var_length(visitor, child):
+    _id = id(child)
+    if _id not in visitor.var_map:
+        interval_var = child.get_associated_interval_var()
+        cpx_interval_var = _get_docplex_interval_var(visitor, interval_var)
+
+        visitor.var_map[_id] = cp.length_of(cpx_interval_var)
+    # There aren't any special types of constraints involving the length, so we
+    # just treat this expression as if it's a normal variable.
+    return False, (_GENERAL, visitor.var_map[_id])
+
+def _before_interval_var_presence(visitor, child):
+    _id = id(child)
+    if _id not in visitor.var_map:
+        interval_var = child.get_associated_interval_var()
+        cpx_interval_var = _get_docplex_interval_var(visitor, interval_var)
+
+        visitor.var_map[_id] = cp.presence_of(cpx_interval_var)
+    # There aren't any special types of constraints involving the presence, so
+    # we just treat this expression as if it's a normal variable.
+    return False, (_GENERAL, visitor.var_map[_id])
+
+def _handle_step_at_node(visitor, node):
+    return cp.step_at(node._time, node._height)
+
+def _handle_step_at_start_node(visitor, node):
+    cpx_var = _get_docplex_interval_var(visitor, node._time)
+    return cp.step_at_start(cpx_var, node._height)
+
+def _handle_step_at_end_node(visitor, node):
+    cpx_var = _get_docplex_interval_var(visitor, node._time)
+    return cp.step_at_end(cpx_var, node._height)
+
+def _handle_pulse_node(visitor, node):
+    cpx_var = _get_docplex_interval_var(visitor, node._interval_var)
+    return cp.pulse(cpx_var, node._height)
+
+_step_function_handles = {
+    StepAt: _handle_step_at_node,
+    StepAtStart: _handle_step_at_start_node,
+    StepAtEnd: _handle_step_at_end_node,
+    Pulse: _handle_pulse_node,
+}
+
+def _handle_negated_step_function_node(visitor, node):
+    return _step_function_handles[node.args[0].__class__](visitor, node.args[0])
+
+_step_function_handles[
+    NegatedStepFunction] = _handle_negated_step_function_node,
```

Comment:
This can all be declared at once (and after `_handle_negated_step_function_node` and `_handle_cumulative_function`)